lightning/src/ln/channelmanager.rs

```diff
@@ -1587,7 +1591,7 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
 	/// Users need to notify the new ChannelManager when a new block is connected or
 	/// disconnected using its `block_connected` and `block_disconnected` methods, starting
 	/// from after `params.latest_hash`.
-	pub fn new(fee_est: F, chain_monitor: M, tx_broadcaster: T, logger: L, keys_manager: K, config: UserConfig, params: ChainParameters) -> Self {
+	pub fn new(fee_est: F, chain_monitor: M, tx_broadcaster: T, router: R, logger: L, keys_manager: K, config: UserConfig, params: ChainParameters) -> Self {
```

Comment:
For the transition, could we drop the `router` parameter and have a different `impl` block for `new` using `DefaultRouter` for `R`?  Then make a test-only constructor if a `TestRouter` is needed?

Reply:
Hmm, that would be great. I guess the `DefaultRouter` could also have a `ProbabilisticScorer`, to avoid an `S` trait param? @TheBlueMatt thoughts on this?

Reply:
Yeah, or alternatively could define a `NullRouter` that implements all the trait methods as `unimplemented!()` if they won't be called yet.

Reply:
Just for the transition. Probably should make a separate `NullRouter` implementation as mentioned above to avoid passing any new parameters as you point out.

Reply:
Sure, could do it all in one release.

I don't think we'd want any migration path if it would result in some `ChannelManager` methods failing when using a `NullRouter`. Better force everyone to upgrade.

Reply:
> I think you'd still be able to use a `NullRouter`, you'd just have to call the manual-route send methods?

Yeah, though I assume it could fail once we move retry logic in. I could be misunderstanding the refactor steps, though.

> Of course I'm not sure that we care about forcing such users to actually manually pass a `NullRouter` to upgrade?

By forcing, I meant by just going ahead and adding the `Router` parameter to `new` in lieu of the `NullRouter` alternative.

Reply:
> > Yeah, though I assume it could fail once we move retry logic in. I could be misunderstanding the refactor steps, though.
> 
> No, I think we'll always want to support sending with a manual path even if you use a `NullRouter`.
> 

Would manually supplied paths be auto-retried if failed? If so, using `NullRouter` would result in retries failing.

> > By forcing, I meant by just going ahead and adding the Router parameter to new in lieu of the NullRouter alternative.
> 
> Hmm, okay, I understood that `NullRouter` here, meant, having a `NullRouter` type that always fails, requiring the user pass a `Router`, but providing the `NullRouter` so that users can pass that when a `Router` is required.

Yeah, my original suggestion for using `NullRouter` would be to keep `ChannelManager::new` untouched while we added support for auto-retries / trampoline. Then the user would largely not need to care that `ChannelManager` had an extra `Router` type parameter. But maybe that isn't the case in practice (e.g., reading a `ChannelManager` from disk).